Userland/Libraries/LibGfx/QOIWriter.cpp

```diff
@@ -0,0 +1,212 @@
+    {
+        if (m_pixels_to_encode == INVALID_PIXELS_TO_ENCODE)
+            return Error::from_string_literal("Trying to encode using a not initialized QOI writer"sv);
+        if (m_pixels_to_encode == 0)
+            return Error::from_string_literal("Trying to encode using a QOI writer that should be finished instead"sv);
+        --m_pixels_to_encode;
+
+        const i8 dr = pixel.red() - get_previous_pixel().red();
+        const i8 dg = pixel.green() - get_previous_pixel().green();
+        const i8 db = pixel.blue() - get_previous_pixel().blue();
+        const i8 da = pixel.alpha() - get_previous_pixel().alpha();
+
+        false
+            || TRY(try_encode_qoi_op_run(dr, dg, db, da))
+            || TRY(try_encode_qoi_op_index(pixel))
+            || TRY(try_encode_qoi_op_diff(dr, dg, db, da))
+            || TRY(try_encode_qoi_op_luma(dr, dg, db, da))
+            || TRY(try_encode_qoi_op_rgb(pixel))
+            || TRY(try_encode_qoi_op_rgba(pixel));
+        set_previous_pixel(pixel);
+        return {};
+    }
+
+    ErrorOr<Stream> finish()
+    {
+        if (m_pixels_to_encode == INVALID_PIXELS_TO_ENCODE)
+            return Error::from_string_literal("Trying to finish a not initialized QOI writer"sv);
+        if (m_pixels_to_encode != 0)
+            return Error::from_string_literal("Prematurely trying to finish a QOI writer"sv);
+
+        m_pixels_to_encode = INVALID_PIXELS_TO_ENCODE;
+        TRY(try_encode_qoi_end_marker());
+        return m_stream;
+    }
+
+private:
+    ErrorOr<void> try_encode_qoi_header(u32 width, u32 height, bool has_alpha_channel)
+    {
+        (void)has_alpha_channel;
+        QOIHeader header {
+            {},
+            AK::convert_between_host_and_big_endian(width),
+            AK::convert_between_host_and_big_endian(height),
+            static_cast<u8>(has_alpha_channel ? 4 : 3),
+            0,
+        };
+        AK::TypedTransfer<u8>::copy(header.magic, QOI_MAGIC, QOI_MAGIC_SIZE);
+        return m_stream.try_append(ReadonlyBytes { &header, sizeof(header) });
+    }
+
+    ErrorOr<bool> try_encode_qoi_op_run(i8 dr, i8 dg, i8 db, i8 da)
+    {
+        const bool is_equal = dr == 0 && dg == 0 && db == 0 && da == 0;
+
+        if (is_equal) {
+            ++m_run;
+            if (m_run < QOI_RUN_MAX && m_pixels_to_encode > 0)
+                return true;
+        } else if (m_run == 0) {
+            return false;
+        }
+
+        VERIFY(qoi_is_valid_run(m_run));
+        // The run-length is stored with a bias of -1.
+        u8 bytes[1] {
+            static_cast<u8>(QOI_OP_RUN | (m_run - 1))
+        };
+        TRY(m_stream.try_append(ReadonlyBytes { bytes }));
+        m_run = 0;
+        return is_equal;
+    }
+
+    ErrorOr<bool> try_encode_qoi_op_index(Color pixel)
+    {
+        u8 index = index_position(pixel);
+        VERIFY(AK::is_in_bounds<u8>(index, 0, 63));
+        if (get_previously_seen_pixel(index) != pixel)
+            return false;
```

Comment:
`ErrorOr<bool>` is normally avoided, as it leads to weird semantics and forces you to handle failures twice, like in your quirky little `||` tree above.  Can we make these `ErrorOr<void>`?